File: frontend/src/types/positions.ts
export interface Position {
  id: number;
  symbol: string;
  exchange: string | null;
  direction: 'LONG' | 'SHORT';
  status: 'open' | 'closed';
  entry_price: number;
  quantity: number;
  entry_time: string;
  entry_signal_id: number | null;

  // Exit config
  stop_loss_price: number | null;
  profit_target_price: number | null;
  stop_loss_pct: number | null;
  profit_target_pct: number | null;
  use_atr_exits: boolean;
  atr_value_at_entry: number | null;
  eod_exit_enabled: boolean;
  max_hold_bars: number | null;

  // Live tracking
  current_price: number | null;
  unrealized_pnl: number | null;
  unrealized_pnl_pct: number | null;
  high_since_entry: number | null;
  low_since_entry: number | null;
  bars_held: number;

  // Exit info (filled when closed)
  exit_price: number | null;
  exit_time: string | null;
  exit_reason: string | null;
  realized_pnl: number | null;
  realized_pnl_pct: number | null;
  realized_pnl_dollar: number | null;

  // Context
  regime_at_entry: string | null;
  regime_at_exit: string | null;
  created_at: string;
}

export interface ExitSignal {
  id: number;
  position_id: number;
  symbol: string;
  exit_type: string;
  urgency: 'low' | 'medium' | 'high' | 'critical';
  trigger_price: number | null;
  current_price: number;
  message: string;
  details: Record<string, unknown> | null;
  acknowledged: boolean;
  acted_on: boolean;
  created_at: string;
}

export interface TradeStats {
  total_trades: number;
  wins: number;
  losses: number;
  win_rate: number | null;
  avg_return_pct: number | null;
  avg_winner_pct: number | null;
  avg_loser_pct: number | null;
  best_trade_pct: number | null;
  worst_trade_pct: number | null;
  profit_factor: number | null;
  avg_bars_held: number | null;
}

export interface TradeInput {
  symbol: string;
  direction: 'LONG' | 'SHORT';
  entry_price: number;
  quantity: number;
  exchange?: string;
  stop_loss_price?: number;
  profit_target_price?: number;
  use_atr_exits?: boolean;
  eod_exit_enabled?: boolean;
}

export interface CloseInput {
  exit_price: number;
  exit_reason?: string;
}

// WebSocket message types
export interface WsExitAlert {
  type: 'exit_alert';
  alert: ExitSignal;
}

export interface WsPositionUpdate {
  type: 'position_update';
  position: Position;
}

export interface WsAck {
  type: 'ack';
  received: string;
}

export interface WsSignalUpdate {
  type: 'signal_update';
  signals: import('./market').Signal[];
  count: number;
  timestamp: string;
}

export type WsMessage = WsExitAlert | WsPositionUpdate | WsAck | WsSignalUpdate;

File: backend/app/schemas/positions.py
from datetime import datetime

from pydantic import BaseModel, Field


class TradeInput(BaseModel):
    """Request to open a new position."""
    symbol: str
    direction: str = Field(pattern="^(LONG|SHORT)$")
    entry_price: float = Field(gt=0)
    quantity: int = Field(gt=0)
    exchange: str | None = None
    entry_signal_id: int | None = None
    entry_time: datetime | None = None

    # Optional exit overrides (system computes defaults if not provided)
    stop_loss_price: float | None = None
    profit_target_price: float | None = None
    stop_loss_pct: float | None = Field(None, ge=0.1, le=20.0)
    profit_target_pct: float | None = Field(None, ge=0.1, le=50.0)
    use_atr_exits: bool = True
    eod_exit_enabled: bool = False
    max_hold_days: int | None = Field(None, ge=1, le=120)  # trading days


class CloseInput(BaseModel):
    """Request to close an open position."""
    exit_price: float = Field(gt=0)
    exit_reason: str | None = "manual"


class ExitUpdateInput(BaseModel):
    """Update exit levels on an open position."""
    stop_loss_price: float | None = None
    profit_target_price: float | None = None
    stop_loss_pct: float | None = Field(None, ge=0.1, le=20.0)
    profit_target_pct: float | None = Field(None, ge=0.1, le=50.0)
    eod_exit_enabled: bool | None = None
    max_hold_days: int | None = Field(None, ge=1, le=120)  # trading days


class PositionEditInput(BaseModel):
    """Edit core position fields (for correcting data entry errors)."""
    entry_price: float | None = Field(None, gt=0)
    quantity: int | None = Field(None, gt=0)
    direction: str | None = Field(None, pattern="^(LONG|SHORT)$")
    stop_loss_price: float | None = None
    profit_target_price: float | None = None


class PositionResponse(BaseModel):
    id: int
    symbol: str
    exchange: str | None = None
    direction: str
    status: str
    entry_price: float
    quantity: int
    entry_time: datetime
    entry_signal_id: int | None = None

    # Exit config
    stop_loss_price: float | None = None
    profit_target_price: float | None = None
    stop_loss_pct: float | None = None
    profit_target_pct: float | None = None
    use_atr_exits: bool = True
    atr_value_at_entry: float | None = None
    eod_exit_enabled: bool = False
    max_hold_days: int | None = None  # trading days

    # Live tracking
    current_price: float | None = None
    unrealized_pnl: float | None = None
    unrealized_pnl_pct: float | None = None
    high_since_entry: float | None = None
    low_since_entry: float | None = None
    bars_held: int = 0

    # Exit info (filled when closed)
    exit_price: float | None = None
    exit_time: datetime | None = None
    exit_reason: str | None = None
    realized_pnl: float | None = None
    realized_pnl_pct: float | None = None
    realized_pnl_dollar: float | None = None

    # Context
    regime_at_entry: str | None = None
    regime_at_exit: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class ExitSignalResponse(BaseModel):
    id: int
    position_id: int
    symbol: str
    exit_type: str
    urgency: str
    trigger_price: float | None = None
    current_price: float
    message: str
    details: dict | None = None
    acknowledged: bool = False
    acted_on: bool = False
    created_at: datetime

    model_config = {"from_attributes": True}


class TradeStatsResponse(BaseModel):
    """Aggregate stats for closed trades."""
    total_trades: int = 0
    wins: int = 0
    losses: int = 0
    win_rate: float | None = None
    avg_return_pct: float | None = None
    avg_winner_pct: float | None = None
    avg_loser_pct: float | None = None
    best_trade_pct: float | None = None
    worst_trade_pct: float | None = None
    profit_factor: float | None = None
    avg_bars_held: float | None = None
